### PR TITLE
[Review] Adds error codes, conventions, 2024-friendly

### PR DESCRIPTION
**Pull request by damirka**
I'll add tests in a follow up

### CODE REVIEW

File: sources/treasury.move [hunk 1]
```diff
@@ -134,126 +172,190 @@ module controlled_treasury::treasury {
...
-        bag::add(&mut treasury.own_capabilities, name, cap);
+        add_cap(treasury, name, cap);
     }
 
-    // Allow the admin to remove capabilities from the treasury
-    // Authorization checks that a capability under the given name is owned by the caller.
-    public fun remove_capability<T : key, C: store + drop>(auth: vector<u8>, treasury: &mut ControlledTreasury<T>, name: vector<u8>, ctx: &mut TxContext) {
+    /// Allow the admin to remove capabilities from the treasury
+    /// Authorization checks that a capability under the given name is owned by the caller.
+    public fun remove_capability<T, C: store + drop>(
```
Comment (manolisliolios) [inline sources/treasury.move]: Should we be defensive here and make sure that there's at least a single admin cap in the Bag?
To avoid the possibility of being locked out of the Treasury?

File: sources/treasury.move [hunk 2]
```diff
@@ -134,126 +172,190 @@ module controlled_treasury::treasury {
...
-    // Allow the admin to add capabilities to the treasury
-    // Authorization checks that a capability under the given name is owned by the caller.
-    public fun add_capability<T : key, C: store>(auth: vector<u8>, treasury: &mut ControlledTreasury<T>, name: vector<u8>, cap: C, ctx: &mut TxContext) {
+    /// Allow the admin to add capabilities to the treasury
+    /// Authorization checks that a capability under the given name is owned by the caller.
+    public fun add_capability<T, C: store + drop>(
+        treasury: &mut ControlledTreasury<T>,
+        auth: vector<u8>,
+        name: vector<u8>,
+        cap: C,
```
Comment (manolisliolios) [inline sources/treasury.move]: Should we be more strict with allowed `C`'s here? 

Or we feel ok allowing any caps to be added to the bag?

Comment (damirka) [reply sources/treasury.move]: They can only be added by Admins, and the requirement `store + drop` is safe for admins to remove them at any moment.

File: sources/treasury.move [hunk 3]
```diff
@@ -134,126 +172,190 @@ module controlled_treasury::treasury {
...
     }
 
-    // Allow the admin to add capabilities to the treasury
-    // Authorization checks that a capability under the given name is owned by the caller.
-    public fun add_capability<T : key, C: store>(auth: vector<u8>, treasury: &mut ControlledTreasury<T>, name: vector<u8>, cap: C, ctx: &mut TxContext) {
+    /// Allow the admin to add capabilities to the treasury
+    /// Authorization checks that a capability under the given name is owned by the caller.
+    public fun add_capability<T, C: store + drop>(
+        treasury: &mut ControlledTreasury<T>,
+        auth: vector<u8>,
```
Comment (manolisliolios) [inline sources/treasury.move]: I do feel like the `auth: vector<u8>` authorization is a bit loose and hard to follow. 

Since we do rely on pre-defined list of Caps (at least for the functionality of the contracts), can't we use the `sender(ctx)` as our `key` for caps? We won't need 2 of the same type of caps for a single address anyways, will we?

File: sources/treasury.move [hunk 4]
```diff
@@ -134,126 +157,153 @@ module controlled_treasury::treasury {
...
         // Check that the amount is within the mint limit
-        // NOTE: Here we can have daily limits by updating the capability within 
+        // NOTE: Here we can have daily limits by updating the capability within
         //       the bag to "remember" totals for the day, for example.
-        assert!(amount <= *limit, 0);
-
-        // Check that the to address is whitelisted, by loading it
-        let whitelist_entry = bag::borrow<vector<u8>, WhitelistEntry>(&treasury.own_capabilities, to);
-        let WhitelistEntry { remote_address } = whitelist_entry;
+        assert!(amount <= *limit, ELimitExceeded);
```
Comment (manolisliolios) [inline sources/treasury.move]: I am worried this example, without implementing an actual daily limit implementation, gives me a bit of a "feeling-safe-but-not-being-safe" vibe, because someone could just call this multiple times within a single PTB.

File: sources/treasury.move [hunk 5]
```diff
@@ -134,126 +157,153 @@ module controlled_treasury::treasury {
...
 
         // Mint and transfer the coins atomically, no holding account of coins
-        event::emit(MintEvent { amount, to: *remote_address });
+        event::emit(MintEvent { amount, to });
```
Comment (manolisliolios) [inline sources/treasury.move]: I think we need `MintEvent<T>` here, otherwise it'll be hard to track off-chain.

File: sources/treasury.move [hunk 6]
```diff
@@ -134,126 +157,153 @@ module controlled_treasury::treasury {
...
+        treasury: &mut ControlledTreasury<T>,
+        coin: Coin<T>,
+        ctx: &mut TxContext
+    ) {
+        assert!(has_cap<T, WhitelistEntry>(treasury, sender(ctx)), ENoAuthRecord);
 
         // Burn the coins atomically, no holding of coins
-        event::emit(BurnEvent { amount: coin::value<T>(&amount), from: *remote_address });
-        coin::burn(&mut treasury.treasury_cap, amount);
+        event::emit(BurnEvent {
```
Comment (manolisliolios) [inline sources/treasury.move]: similar to `MintEvent`

File: sources/treasury.move [hunk 7]
```diff
@@ -134,126 +157,153 @@ module controlled_treasury::treasury {
...
+    public fun remove_deny_address<T>(
+        treasury: &mut ControlledTreasury<T>,
+        deny_list: &mut DenyList,
+        entry: address,
+        ctx: &mut TxContext
+    ) {
+        assert!(has_cap<T, DenyMutCap>(treasury, sender(ctx)), ENoAuthRecord);
+        assert!(coin::deny_list_contains<T>(deny_list, entry), ENoDenyEntry);
 
         // Add to deny list
```
Comment (manolisliolios) [inline sources/treasury.move]: Comment typo `// remove from denylist`

File: sources/treasury.move [hunk 8]
```diff
@@ -83,49 +88,67 @@ module controlled_treasury::treasury {
...
-        let cap = ControlledTreasury {
+    /// Create a new controlled treasury by wrapping the treasury cap of a coin
+    /// Th treasure becomes a public object with an initial Admin capbility assigned to
+    /// the owner of the treasury as provided.
+    ///
+    /// The `ControlledTreasury` has to be `share`d after the creation.
+    public fun new<T>(
+        treasury_cap: TreasuryCap<T>,
+        deny_cap: DenyCap<T>,
+        owner: address,
```
Comment (manolisliolios) [inline sources/treasury.move]: What's the motivation for having a separate `owner` than the tx sender?

File: sources/treasury.move (hunk 5 above)
Comment (damirka) [reply sources/treasury.move]: Took me a while to figure out what you meant here...

File: sources/treasury.move (hunk 8 above)
Comment (damirka) [reply sources/treasury.move]: That's a good one, and probably for @gdanezis

File: sources/treasury.move (hunk 4 above)
Comment (damirka) [reply sources/treasury.move]: Absolutely but I also don't understand the requirements for the feature. Should we disable it for now?

Comment (manolisliolios) [reply sources/treasury.move]: Yeah, probably removing is best!

Or, adding a `per-epoch` limitter instead which is simple to do and has some actual limitting logic involved.

File: sources/treasury.move [hunk 9]
```diff
@@ -1,131 +1,147 @@
...
+    public fun new_deny_mut_cap(): DenyMutCap { DenyMutCap {} }
 
-    // Create a new controlled treasury by wrapping the treasury cap of a coin
-    // Th treasure becomes a public object with an initial Admin capbility assigned to 
-    // the owner of the treasury as provided.
-    public fun new<T : key>(treasury_cap: TreasuryCap<T>, deny_cap: DenyCap<T>, owner: address, ctx: &mut TxContext) {
-        let cap = ControlledTreasury {
+    /// Create a new controlled treasury by wrapping the treasury cap of a coin
+    /// Th treasure becomes a public object with an initial Admin capbility assigned to
+    /// the owner of the treasury as provided.
```
Comment (manolisliolios) [inline sources/treasury.move]:
```suggestion
    /// The `ControlledTreasury` becomes a public shared object with an initial Admin capability assigned to
    /// the provided owner.
```

File: sources/treasury.move [hunk 10]
```diff
@@ -134,126 +150,188 @@ module controlled_treasury::treasury {
...
 
     // Allow an authorized multi-sig to mint and transfer coins to a whitelisted address
-    public fun mint_and_transfer<T : key>(auth: vector<u8>, treasury: &mut ControlledTreasury<T>, to: vector<u8>, amount: u64, ctx: &mut TxContext) {
+    ///
+    /// Aborts if:
+    /// - sender does not have MintCap assigned to them
+    /// - the amount is higher than the defined limit on MintCap
+    /// - the receiver is not Whitelisted
+    ///
+    // Emits: MintEvent
```
Comment (manolisliolios) [inline sources/treasury.move]:
```suggestion
    /// Emits: MintEvent
```

File: sources/treasury.move (hunk 8 above)
Comment (gdanezis) [reply sources/treasury.move]: This is in case you are creating a coin for someone else? Adds flexibility. In particular you may want the first admin to be a multisig but you may create the treasury using a normal sig?

**Review by manolisliolios — APPROVED** [review]
LGTM!

File: sources/treasury.move (hunk 8 above)
Comment (manolisliolios) [reply sources/treasury.move]: I see! I do not see any problems with it, I was just going on a paranoid review mode questioning everything!
